vllm/attention/backends/mla/utils.py

```diff
@@ -199,9 +204,14 @@ def _v_up_proj_and_o_proj(self, x):
                 output = output_parallel
             return output
         else:
-            x = torch.einsum("bnl,lnv->bnv", x, self.W_UV)
-            return self.o_proj(x.reshape(-1,
-                                         self.num_heads * self.v_head_dim))[0]
+            # chendi: this is a cherry-pick of missing commit from upstream
+            # Convert from (B, N, L) to (N, B, L)
+            x = x.view(-1, self.num_heads, self.kv_lora_rank).transpose(0, 1)
+            # Multiply (N, B, L) x (N, L, V) -> (N, B, V)
+            x = torch.bmm(x, self.W_UV)
+            # Convert from (N, B, V) to (B, N * V)
+            x = x.transpose(0, 1).reshape(-1, self.num_heads * self.v_head_dim)
+            return self.o_proj(x)[0]
 
     def _q_proj_and_k_up_proj(self, x):
         if envs.VLLM_MLA_PERFORM_MATRIX_ABSORPTION:
```

Comment:
can we remove this since upstream also removed?

Reply:
We noticed accuracy drop with VLLM_MLA_PERFORM_MATRIX_ABSORPTION=0, let's keep the option to enable it before the accuracy issues is solved.